### PR TITLE
Merge ToolConfWatchers and separate from Toolbox

### PR DESCRIPTION
**Pull request by mvdbeek**
This should avoid some complexities associated with managing threads.

### CODE REVIEW

File: lib/galaxy/tools/toolbox/watcher.py [hunk 1]
```diff
@@ -120,23 +163,22 @@ def check(self):
                 if removed_ids:
                     do_reload = True
             if do_reload:
-                with self._lock:
-                    t = threading.Thread(target=self.event_handler.on_any_event)
-                    t.daemon = True
-                    t.start()
+                    self.reload_callback()
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/watcher.py]: You can reduce the indentation here.

**Comment by nsoranzo** [comment]
Looks much cleaner, thanks @mvdbeek!

File: lib/galaxy/tools/toolbox/watcher.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/tools/toolbox/watcher.py]: I'm surprised that wasn't flagged by tox!

Comment (nsoranzo) [reply lib/galaxy/tools/toolbox/watcher.py]: Looks like I'm more :eagle: than flake8 :laughing:

File: lib/galaxy/tools/toolbox/watcher.py [hunk 2]
```diff
@@ -14,6 +14,10 @@
     PollingObserver = None
     can_watch = False
 
+from galaxy.queue_worker import (
+    reload_data_managers,
+    reload_toolbox,
+)
```
Comment (jmchilton) [inline lib/galaxy/tools/toolbox/watcher.py]: Nothing in ``galaxy.tools.toolbox`` should depend on ``galaxy.queue_worker`` IMO. The former module is meant to stand-alone from the app and models - hence the subclassing of ToolBox from AbstractToolBox and the inclusion on that module in galaxy-lib. Can you refactor this config watcher stuff into say ``galaxy.tools.watchers`` or ``galaxy.webapps.galaxy.config_watchers`` perhaps.

Comment (mvdbeek) [reply lib/galaxy/tools/toolbox/watcher.py]: OK, I think I've done that in 1a5e337420e4832593b6e30089c0022291fd6078.

**Comment by mvdbeek** [comment]
This is ready for review now.
I think this is good progress in itself, I will try to further merge the individual threads into 1, if I can speed up the data table reloading.

File: lib/galaxy/webapps/galaxy/config_watchers.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    reload_toolbox,
+)
+from galaxy.tools.toolbox.watcher import (
+    get_tool_conf_watcher,
+    get_tool_data_dir_watcher
+)
+
+
+class ConfigWatchers(object):
+    """Contains ToolConfWatcher, ToolWatcher and ToolDataWatcher objects."""
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/config_watchers.py:13]: It seems this class does not contain a `ToolWatcher`.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/config_watchers.py:13]: Right, I've made that change in f79fb90

**Comment by jmchilton** [comment]
Agreed with @nsoranzo - this is really nice! What a better separation of concerns - very crisp. 💯

**Pull request merged**